src/runtime/type.ts

```diff
@@ -381,42 +381,42 @@ export class TypeVariable extends Type {
 }
 
 
-export class CollectionType extends Type {
-  readonly collectionType: Type;
+export class CollectionType<T extends Type> extends Type {
+  readonly collectionType: T;
 
-  constructor(collectionType: Type) {
+  constructor(collectionType: T) {
```

Comment:
can this also be `T extends Type` ?

Reply:
Adding that does not compile. I believe that the `extends Type` is inferred by the TS compiler.